src/Archive.hs

```diff
@@ -0,0 +1,55 @@
+module Archive
+    ( putOutput
+    , getOutput
+    , deleteOutput
+    ) where
+
+import Import
+
+import qualified Aws as Aws
+import qualified Aws.S3 as S3
+
+putOutput :: Token -> [Output] -> Handler ()
+putOutput token outputs = do
+    bucket <- appS3Bucket . appSettings <$> getYesod
+
+    let bytestrings = map (encodeUtf8 . outputContent) outputs
+        contentLength = fromIntegral $ sum $ map length bytestrings
+
+    $(logDebug) $ "archiving output to S3 " <> tokenText token
+
+    runS3 (S3.putObject bucket (tokenText token)
+        $ requestBodySource contentLength $ yieldMany bytestrings)
+        $ ignore
+
+getOutput :: Token -> Sink ByteString (ResourceT IO) a -> Handler a
+getOutput token sink = do
+    bucket <- appS3Bucket . appSettings <$> getYesod
+
+    $(logDebug) $ "retrieving archived output " <> tokenText token
+
+    runS3 (S3.getObject bucket $ tokenText token)
+        $ \rsp -> responseBody (S3.gorResponse rsp) $$+- sink
+
+deleteOutput :: Token -> Handler ()
+deleteOutput token = do
+    bucket <- appS3Bucket . appSettings <$> getYesod
+
+    $(logDebug) $ "deleting archived output " <> tokenText token
+
+    runS3 (S3.deleteObjects bucket [tokenText token]) $ ignore
```

Comment:
right, `runS3` takes a function ... ignore that last comment ... however, do you need apply here `$`?